bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableIO.java

```diff
@@ -692,7 +699,34 @@ void initializeScanner() throws IOException {
     /** Calls {@link ResultScanner#next()}. */
     @Override
     public boolean advance() throws IOException {
+      try {
+        return tryAdvance();
+      } catch (Throwable e) {
+        // if retry idle timeout is disabled, throw the exception
+        if (!source.getConfiguration().toHBaseConfig().getBoolean(RETRY_IDLE_TIMEOUT, true)) {
+          throw e;
+        }
+        Throwable exception = e;
+        while (exception != null && !(exception instanceof WatchdogTimeoutException)) {
+          exception = exception.getCause();
+        }
+        if (exception != null) {
+          READER_LOG.warn("got idle timeout exception, try resetting the scanner", e);
+          if (attempt.decrementAndGet() >= 0
+              && exception.getMessage() != null
+              && exception.getMessage().contains("idle")) {
```

Comment:
isnt this backwards? Shouldn't you check for idle in the message before logging that you got an idle timeout?

Reply:
I think it would be cleaner as a series of early returns:

```
if (!config.getBoolean(RETRY_IDLE_TIMEOUT, true)) {
  throw e;
}

Throwable t = Utils.findCause(WatchdogTimeoutException.class);
if (t == null) {
  throw e;
}

if (t.getMessage() == null || !t.getMessage().contains("idle")) {
   throw e;
}
if (attempt.decrementAndGet() <= 0) {
   throw e;
}
resetScanner();
return tryAdvance();
```